### Exploratory Data Analysis
Understanding the dataset to explore how the data is present in the database and if there is a need of creating some aggregated table that can help with:

- Vendor selection for profitability
- Product Pricing Optimization

In [ ]:
import pandas as pd
from sqlalchemy import create_engine
from dotenv import load_dotenv
import os

load_dotenv()

True

In [7]:
conn = create_engine(f"postgresql+psycopg2://{os.getenv('DB_USERNAME')}:{os.getenv('DB_PASSWORD')}@{os.getenv('DB_HOST')}:{os.getenv('DB_PORT')}/{os.getenv('DB_NAME')}")


In [8]:
tables = pd.read_sql("SELECT table_name FROM information_schema.tables WHERE table_schema='public'", conn)
tables

table_name
0  begin_inventory
1  purchase_prices
2    end_inventory
3        purchases
4            sales
5   vendor_invoice

In [9]:
for table in tables['table_name']:
    print('-'*50, f'{table}:','-'*50)
    print('Count of records:',pd.read_sql(f'SELECT count(*) as count FROM {table}', conn)['count'][0])
    

-------------------------------------------------- begin_inventory: --------------------------------------------------
Count of records: 206529
-------------------------------------------------- purchase_prices: --------------------------------------------------
Count of records: 12261
-------------------------------------------------- end_inventory: --------------------------------------------------
Count of records: 224489
-------------------------------------------------- purchases: --------------------------------------------------
Count of records: 2372474
-------------------------------------------------- sales: --------------------------------------------------
Count of records: 12825363
-------------------------------------------------- vendor_invoice: --------------------------------------------------
Count of records: 5543


In [10]:
for table in tables['table_name']:
    print(f"{'--'*10} First 5 rows of table: {table} {'--'*10}")
    display(pd.read_sql(f"SELECT * FROM {table} LIMIT 5" , conn))


-------------------- First 5 rows of table: begin_inventory --------------------


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   
2  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
3  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
4  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2024-01-01  
1  750mL       7  10.99  2024-01-01  
2  750mL       6  36.99  2024-01-01  
3  750mL       3  38.99  2024-01-01  
4  750mL       6  34.99  2024-01-01

-------------------- First 5 rows of table: purchase_prices --------------------


Brand                  Description  Price   Size Volume  Classification  \
0     58  Gekkeikan Black & Gold Sake  12.99  750mL    750               1   
1     62     Herradura Silver Tequila  36.99  750mL    750               1   
2     63   Herradura Reposado Tequila  38.99  750mL    750               1   
3     72         No. 3 London Dry Gin  34.99  750mL    750               1   
4     75    Three Olives Tomato Vodka  14.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0           9.28          8320  SHAW ROSS INT L IMP LTD      
1          28.67          1128  BROWN-FORMAN CORP            
2          30.46          1128  BROWN-FORMAN CORP            
3          26.11          9165  ULTRA BEVERAGE COMPANY LLP   
4          10.94          7245  PROXIMO SPIRITS INC.

-------------------- First 5 rows of table: end_inventory --------------------


InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   
2  1_HARDERSFIELD_63      1  HARDERSFIELD     63   Herradura Reposado Tequila   
3  1_HARDERSFIELD_72      1  HARDERSFIELD     72         No. 3 London Dry Gin   
4  1_HARDERSFIELD_75      1  HARDERSFIELD     75    Three Olives Tomato Vodka   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2024-12-31  
1  750mL       7  36.99  2024-12-31  
2  750mL       7  38.99  2024-12-31  
3  750mL       4  34.99  2024-12-31  
4  750mL       7  14.99  2024-12-31

-------------------- First 5 rows of table: purchases --------------------


InventoryId  Store  Brand                   Description   Size  \
0    69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1     30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   
2    34_PITMERDEN_5215     34   5215  TGI Fridays Long Island Iced  1.75L   
3  1_HARDERSFIELD_5255      1   5255  TGI Fridays Ultimte Mudslide  1.75L   
4    76_DONCASTER_2034     76   2034     Glendalough Double Barrel  750mL   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2023-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
2          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
3          4466  AMERICAN VINTAGE BEVERAGE        8137  2023-12-22   
4           388  ATLANTIC IMPORTING COMPANY       8169  2023-12-24   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2024-01-02  2024-01-04  2024-02-16          35.71         6   214.26   
1    2024-01-01  2024-01-07  2024-02-21           9.35         4    37.40   
2    2024-01-02  2024-01-07  2024-02-21           9.41         5    47.05   
3    2024-01-01  2024-01-07  2024-02-21           9.35         6    56.10   
4    2024-01-02  2024-01-09  2024-02-16          21.32         5   106.60   

   Classification  
0               1  
1               1  
2               1  
3               1  
4               1

-------------------- First 5 rows of table: sales --------------------


InventoryId  Store  Brand                 Description        Size  \
0  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
1  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
2  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
3  1_HARDERSFIELD_1004      1   1004  Jim Beam w/2 Rocks Glasses       750mL   
4  1_HARDERSFIELD_1005      1   1005     Maker's Mark Combo Pack  375mL 2 Pk   

   SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0              1         16.49       16.49  2024-01-01   750.0   
1              2         32.98       16.49  2024-01-02   750.0   
2              1         16.49       16.49  2024-01-03   750.0   
3              1         14.49       14.49  2024-01-08   750.0   
4              2         69.98       34.99  2024-01-09   375.0   

   Classification  ExciseTax  VendorNo                   VendorName  
0               1       0.79     12546  JIM BEAM BRANDS COMPANY      
1               1       1.57     12546  JIM BEAM BRANDS COMPANY      
2               1       0.79     12546  JIM BEAM BRANDS COMPANY      
3               1       0.79     12546  JIM BEAM BRANDS COMPANY      
4               1       0.79     12546  JIM BEAM BRANDS COMPANY

-------------------- First 5 rows of table: vendor_invoice --------------------


VendorNumber                   VendorName InvoiceDate  PONumber  \
0           105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1          4466  AMERICAN VINTAGE BEVERAGE    2024-01-07      8137   
2           388  ATLANTIC IMPORTING COMPANY   2024-01-09      8169   
3           480  BACARDI USA INC              2024-01-12      8106   
4           516  BANFI PRODUCTS CORP          2024-01-07      8170   

       PODate     PayDate  Quantity    Dollars  Freight Approval  
0  2023-12-21  2024-02-16         6     214.26     3.47     None  
1  2023-12-22  2024-02-21        15     140.55     8.57     None  
2  2023-12-24  2024-02-16         5     106.60     4.61     None  
3  2023-12-20  2024-02-05     10100  137483.78  2935.20     None  
4  2023-12-24  2024-02-12      1935   15527.25   429.20     None

### Insight :
1. This data is of year 2024 __Begin_inventory__ contain data of product starting of this year and __end_inventory__ contain data of product the remain in inventory at end of year 

In [11]:
purchases = pd.read_sql_query('Select * from purchases where "VendorNumber" = 105', conn)
purchases

InventoryId  Store  Brand                Description   Size  \
0      69_MOUNTMEND_8412     69   8412  Tequila Ocho Plata Fresno  750mL   
1     66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
2       15_WANBORNE_8412     15   8412  Tequila Ocho Plata Fresno  750mL   
3     66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
4        10_HORNSEY_8412     10   8412  Tequila Ocho Plata Fresno  750mL   
5      73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
6      76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
7      76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
8      50_MOUNTMEND_8412     50   8412  Tequila Ocho Plata Fresno  750mL   
9      73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
10   37_PAETHSMOUTH_8412     37   8412  Tequila Ocho Plata Fresno  750mL   
11    66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
12      41_LARNWICK_8412     41   8412  Tequila Ocho Plata Fresno  750mL   
13     34_PITMERDEN_8412     34   8412  Tequila Ocho Plata Fresno  750mL   
14     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
15    66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
16     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
17     50_MOUNTMEND_8412     50   8412  Tequila Ocho Plata Fresno  750mL   
18     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
19     38_GOULCREST_8412     38   8412  Tequila Ocho Plata Fresno  750mL   
20    66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
21     69_MOUNTMEND_8412     69   8412  Tequila Ocho Plata Fresno  750mL   
22    66_EANVERNESS_8412     66   8412  Tequila Ocho Plata Fresno  750mL   
23      15_WANBORNE_8412     15   8412  Tequila Ocho Plata Fresno  750mL   
24       54_GUTHRAM_2529     54   2529                  Right Gin  750mL   
25      15_WANBORNE_8412     15   8412  Tequila Ocho Plata Fresno  750mL   
26  47_PELLA'S WISH_8412     47   8412  Tequila Ocho Plata Fresno  750mL   
27     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
28     73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
29   37_PAETHSMOUTH_8412     37   8412  Tequila Ocho Plata Fresno  750mL   
30       60_IRRAGIN_8412     60   8412  Tequila Ocho Plata Fresno  750mL   
31     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
32     55_DRY GULCH_8412     55   8412  Tequila Ocho Plata Fresno  750mL   
33      41_LARNWICK_8412     41   8412  Tequila Ocho Plata Fresno  750mL   
34     73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
35     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fresno  750mL   
36     34_PITMERDEN_8412     34   8412  Tequila Ocho Plata Fresno  750mL   
37     32_MOUNTMEND_8412     32   8412  Tequila Ocho Plata Fresno  750mL   
38     32_MOUNTMEND_8412     32   8412  Tequila Ocho Plata Fresno  750mL   
39    67_EANVERNESS_8412     67   8412  Tequila Ocho Plata Fresno  750mL   
40     38_GOULCREST_8412     38   8412  Tequila Ocho Plata Fresno  750mL   
41       54_GUTHRAM_2529     54   2529                  Right Gin  750mL   
42    74_PAENTMARWY_8412     74   8412  Tequila Ocho Plata Fresno  750mL   
43       60_IRRAGIN_8412     60   8412  Tequila Ocho Plata Fresno  750mL   
44     38_GOULCREST_8412     38   8412  Tequila Ocho Plata Fresno  750mL   
45     50_MOUNTMEND_8412     50   8412  Tequila Ocho Plata Fresno  750mL   
46    67_EANVERNESS_8412     67   8412  Tequila Ocho Plata Fresno  750mL   
47     73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
48     38_GOULCREST_8412     38   8412  Tequila Ocho Plata Fresno  750mL   
49     73_DONCASTER_8412     73   8412  Tequila Ocho Plata Fresno  750mL   
50      49_GARIGILL_8412     49   8412  Tequila Ocho Plata Fresno  750mL   
51     76_DONCASTER_8412     76   8412  Tequila Ocho Plata Fr

In [12]:
purchase_prices = pd.read_sql_query('Select * from purchase_prices where "VendorNumber" = 105', conn)
purchase_prices

Brand                  Description  Price   Size Volume  Classification  \
0   8412    Tequila Ocho Plata Fresno  49.99  750mL    750               1   
1   8419  Ocho Tequila Plata Corrales   6.99  100mL    100               1   
2   2529                    Right Gin  29.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0          35.71           105  ALTAMAR BRANDS LLC           
1           5.30           105  ALTAMAR BRANDS LLC           
2          23.25           105  ALTAMAR BRANDS LLC

In [13]:
vendor_invoice = pd.read_sql_query('Select * from vendor_invoice where "VendorNumber" = 105', conn)
vendor_invoice

VendorNumber                   VendorName InvoiceDate  PONumber  \
0            105  ALTAMAR BRANDS LLC           2024-01-04      8124   
1            105  ALTAMAR BRANDS LLC           2024-01-17      8234   
2            105  ALTAMAR BRANDS LLC           2024-02-22      8836   
3            105  ALTAMAR BRANDS LLC           2024-03-05      8979   
4            105  ALTAMAR BRANDS LLC           2024-03-13      9016   
5            105  ALTAMAR BRANDS LLC           2024-03-14      9131   
6            105  ALTAMAR BRANDS LLC           2024-04-09      9449   
7            105  ALTAMAR BRANDS LLC           2024-04-12      9518   
8            105  ALTAMAR BRANDS LLC           2024-04-28      9719   
9            105  ALTAMAR BRANDS LLC           2024-05-10      9871   
10           105  ALTAMAR BRANDS LLC           2024-05-16      9952   
11           105  ALTAMAR BRANDS LLC           2024-05-23     10021   
12           105  ALTAMAR BRANDS LLC           2024-06-04     10222   
13           105  ALTAMAR BRANDS LLC           2024-06-12     10377   
14           105  ALTAMAR BRANDS LLC           2024-06-21     10465   
15           105  ALTAMAR BRANDS LLC           2024-06-28     10578   
16           105  ALTAMAR BRANDS LLC           2024-06-29     10632   
17           105  ALTAMAR BRANDS LLC           2024-07-07     10797   
18           105  ALTAMAR BRANDS LLC           2024-07-08     10835   
19           105  ALTAMAR BRANDS LLC           2024-07-11     10998   
20           105  ALTAMAR BRANDS LLC           2024-07-29     11170   
21           105  ALTAMAR BRANDS LLC           2024-08-04     11310   
22           105  ALTAMAR BRANDS LLC           2024-08-10     11381   
23           105  ALTAMAR BRANDS LLC           2024-08-23     11565   
24           105  ALTAMAR BRANDS LLC           2024-08-31     11666   
25           105  ALTAMAR BRANDS LLC           2024-09-10     11817   
26           105  ALTAMAR BRANDS LLC           2024-09-15     11835   
27           105  ALTAMAR BRANDS LLC           2024-09-27     11938   
28           105  ALTAMAR BRANDS LLC           2024-10-03     12057   
29           105  ALTAMAR BRANDS LLC           2024-10-11     12376   
30           105  ALTAMAR BRANDS LLC           2024-10-23     12486   
31           105  ALTAMAR BRANDS LLC           2024-11-01     12537   
32           105  ALTAMAR BRANDS LLC           2024-10-31     12634   
33           105  ALTAMAR BRANDS LLC           2024-11-11     12765   
34           105  ALTAMAR BRANDS LLC           2024-11-20     12871   
35           105  ALTAMAR BRANDS LLC           2024-11-28     13053   
36           105  ALTAMAR BRANDS LLC           2024-12-19     13293   
37           105  ALTAMAR BRANDS LLC           2024-12-20     13349   
38           105  ALTAMAR BRANDS LLC           2025-01-03     13464   
39           105  ALTAMAR BRANDS LLC           2025-01-04     13575   

        PODate     PayDate  Quantity  Dollars  Freight Approval  
0   2023-12-21  2024-02-16         6   214.26     3.47     None  
1   2023-12-29  2024-02-22        12   428.52     2.19     None  
2   2024-02-09  2024-03-31        18   642.78     3.02     None  
3   2024-02-18  2024-04-15         6   214.26     1.11     None  
4   2024-02-21  2024-04-11         6   214.26     1.09     None  
5   2024-02-29  2024-04-17         3   107.13     0.57     None  
6   2024-03-22  2024-05-12         6   214.26     1.07     None  
7   2024-03-27  2024-05-17         6   214.26     1.11     None  
8   2024-04-10  2024-06-12        12   428.52     2.31     None  
9   2024-04-20  2024-06-10         6   214.26     1.18     None  
10  2024-04-26  2024-06-25        18   568.02     3.12     None  
11  2024-05-01  2024-06-26         3   107.13     0.57     None  
12  2024-05-15  2024-07-05         2    71.42     0.39     None  
13  2024-05-25  2024-07-19         6   214.26     1.16     None  
14  2024-05-31  2024-07-16        12   428.52     2.36     None  
15  2024-06-08  2024-08-02         6

In [14]:
sales = pd.read_sql_query('Select * from sales where "VendorNo" = 105', conn)
sales

InventoryId  Store  Brand                  Description   Size  \
0    1_HARDERSFIELD_8412      1   8412    Tequila Ocho Plata Fresno  750mL   
1    1_HARDERSFIELD_8419      1   8419  Ocho Tequila Plata Corrales  100mL   
2    1_HARDERSFIELD_8419      1   8419  Ocho Tequila Plata Corrales  100mL   
3    1_HARDERSFIELD_8419      1   8419  Ocho Tequila Plata Corrales  100mL   
4    1_HARDERSFIELD_8419      1   8419  Ocho Tequila Plata Corrales  100mL   
..                   ...    ...    ...                          ...    ...   
262    76_DONCASTER_8412     76   8412    Tequila Ocho Plata Fresno  750mL   
263    76_DONCASTER_8412     76   8412    Tequila Ocho Plata Fresno  750mL   
264      10_HORNSEY_8412     10   8412    Tequila Ocho Plata Fresno  750mL   
265     49_GARIGILL_8412     49   8412    Tequila Ocho Plata Fresno  750mL   
266   67_EANVERNESS_8412     67   8412    Tequila Ocho Plata Fresno  750mL   

     SalesQuantity  SalesDollars  SalesPrice   SalesDate  Volume  \
0                1         49.99       49.99  2024-01-24   750.0   
1                1          6.99        6.99  2024-01-08   100.0   
2                1          6.99        6.99  2024-01-12   100.0   
3                1          6.99        6.99  2024-01-17   100.0   
4                1          6.99        6.99  2024-01-19   100.0   
..             ...           ...         ...         ...     ...   
262              1         49.99       49.99  2024-11-22   750.0   
263              1         49.99       49.99  2024-11-30   750.0   
264              1         49.99       49.99  2024-12-27   750.0   
265              1         49.99       49.99  2024-12-02   750.0   
266              1         49.99       49.99  2024-12-23   750.0   

     Classification  ExciseTax  VendorNo                   VendorName  
0                 1       0.79       105  ALTAMAR BRANDS LLC           
1                 1       0.10       105  ALTAMAR BRANDS LLC           
2                 1       0.10       105  ALTAMAR BRANDS LLC           
3                 1       0.10       105  ALTAMAR BRANDS LLC           
4                 1       0.10       105  ALTAMAR BRANDS LLC           
..              ...        ...       ...                          ...  
262               1       0.79       105  ALTAMAR BRANDS LLC           
263               1       0.79       105  ALTAMAR BRANDS LLC           
264               1       0.79       105  ALTAMAR BRANDS LLC           
265               1       0.79       105  ALTAMAR BRANDS LLC           
266               1       0.79       105  ALTAMAR BRANDS LLC           

[267 rows x 14 columns]

In [15]:
purchases.groupby(['Brand','PurchasePrice'])[['Quantity','Dollars']].sum()


Quantity  Dollars
Brand PurchasePrice                   
2529  23.25                12    279.0
8412  35.71               320  11427.2

In [16]:
purchase_prices

Brand                  Description  Price   Size Volume  Classification  \
0   8412    Tequila Ocho Plata Fresno  49.99  750mL    750               1   
1   8419  Ocho Tequila Plata Corrales   6.99  100mL    100               1   
2   2529                    Right Gin  29.99  750mL    750               1   

   PurchasePrice  VendorNumber                   VendorName  
0          35.71           105  ALTAMAR BRANDS LLC           
1           5.30           105  ALTAMAR BRANDS LLC           
2          23.25           105  ALTAMAR BRANDS LLC

**Notes:** 
1. VendorNumber 105 offer 3 brand but only two of them are purchases by the Platform i.e Brand(8412,2529) but Brand 8419 haven't actually bought any of it yet.
2. No such record find for vendorNumber 4466,388
3. VendorNumber 480 offer 184 brands but only 165 are purchased by the Platform. 19 brands are not purchased by the Platform.
4. VendorNumber 516 offer 149 brands but only 131 are purchased by the Platform. 18 brands are not purchased by the Platform.

In [17]:
sales.groupby(['Brand'])[['SalesQuantity','SalesPrice','SalesDollars']].sum()

SalesQuantity  SalesPrice  SalesDollars
Brand                                         
2529              12       59.98        359.88
8412             307    12947.41      15346.93
8419               6       41.94         41.94

**Notes**
1. For VendorNumber 105 Most selling brand is 8412 and least selling brand is 8419.

**Notes** 
- `Purchase` table contain data that are purhcased by platform in which PurchasePrice shows at how much it is purchased from vendor and `purchase_price` table price show at what price vendor is going to sell product to customer.
- Similarly in `sales` table SalesPrice = This is the price of a single unit of the product that customer pay to platform. `SalesDollars = SalesPrice × SalesQuantity`
- The purchases table contains actual purchase data, including the date of purchase, products (brands) purchased by vendors, the amount paid (in dollars), and the quantity purchased.
- The purchase price column is derived from the purchase_prices table, which provides product-wise actual and purchase prices. The combination of vendor and brand is unique in this table.
- The vendor_invoice table aggregates data from the purchases table, summarizing quantity and dollar amounts, along with an additional column for freight.
- This table maintains uniqueness based on vendor and PO number.
- The sales table captures actual sales transactions, detailing the brands purchased by vendors, the quantity sold, the selling price, and the revenue earned.
----
As the data that we need for analysis is distributed in different tables, we need to create a summary table containing:
- purchase transactions made by vendors
- sales transaction data
- freight costs for each vendor
- actual product prices from vendors

In [18]:
## freight costs for each vendor
query = 'SELECT "VendorNumber","VendorName", SUM("Freight") AS "FreightCost" FROM vendor_invoice GROUP BY "VendorNumber","VendorName"'
freight_summary = pd.read_sql_query(query, conn)
freight_summary

VendorNumber                   VendorName  FreightCost
0           90053  INCREDIBREW INC                   124.44
1            4425        MARTIGNETTI COMPANIES    144719.92
2            1587  VINEYARD BRANDS LLC                54.23
3           10000  MAJESTIC FINE WINES             17587.59
4           11567  CAMPARI AMERICA                 20964.81
..            ...                          ...          ...
124          9552  M S WALKER INC                  55551.82
125          3924  HEAVEN HILL DISTILLERIES        14069.87
126          1587  VINEYARD BRANDS INC              6015.86
127         90032  LABELLE VYDS AND WINERY           778.25
128         90037  THE PIERPONT GROUP LLC             27.98

[129 rows x 3 columns]

In [19]:
purchase_prices.columns

Index(['Brand', 'Description', 'Price', 'Size', 'Volume', 'Classification',
       'PurchasePrice', 'VendorNumber', 'VendorName'],
      dtype='str')

In [20]:
purchases.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'VendorNumber',
       'VendorName', 'PONumber', 'PODate', 'ReceivingDate', 'InvoiceDate',
       'PayDate', 'PurchasePrice', 'Quantity', 'Dollars', 'Classification'],
      dtype='str')

In [21]:
query = """
SELECT 
    p."Brand",
    p."VendorNumber", 
    p."VendorName", 
    p."PurchasePrice", 
    pp."Volume", 
    pp."Price" AS "ActualPrice", 
    SUM(p."Quantity") AS "TotalPurchaseQuantity", 
    SUM(p."Dollars") AS "TotalPurchaseDollars" 
FROM purchases p 
JOIN purchase_prices pp ON p."Brand" = pp."Brand" 
WHERE p."PurchasePrice" > 0
GROUP BY 
    p."Brand",
    p."VendorNumber", 
    p."VendorName",  
    p."PurchasePrice",
    pp."Volume",
    pp."Price"
ORDER BY "TotalPurchaseDollars"
"""
purchase_summary = pd.read_sql_query(query, conn)
purchase_summary

Brand  VendorNumber                   VendorName  PurchasePrice Volume  \
0       3065          7245  PROXIMO SPIRITS INC.                  0.71     50   
1       6127          3960  DIAGEO NORTH AMERICA INC              1.47    200   
2       9123          3924  HEAVEN HILL DISTILLERIES              0.74     50   
3       5683          8004  SAZERAC CO INC                        0.39     50   
4       8527          9815  WINE GROUP INC                        1.32    750   
...      ...           ...                          ...            ...    ...   
10687   3545          3960  DIAGEO NORTH AMERICA INC             21.89   1750   
10688   4261          3960  DIAGEO NORTH AMERICA INC             16.17   1750   
10689   8068         17035  PERNOD RICARD USA                    18.24   1750   
10690   3405          4425        MARTIGNETTI COMPANIES          23.19   1750   
10691   1233          1128  BROWN-FORMAN CORP                    26.27   1750   

       ActualPrice  TotalPurchaseQuantity  TotalPurchaseDollars  
0             0.99                    1.0                  0.71  
1             1.99                    1.0                  1.47  
2             0.99                    2.0                  1.48  
3             0.49                    6.0                  2.34  
4             4.99                    2.0                  2.64  
...            ...                    ...                   ...  
10687        29.99               138109.0            3023206.01  
10688        22.99               201682.0            3261197.94  
10689        24.99               187407.0            3418303.68  
10690        28.99               164038.0            3804041.22  
10691        36.99               145080.0            3811251.60  

[10692 rows x 8 columns]

In [22]:
sales.columns

Index(['InventoryId', 'Store', 'Brand', 'Description', 'Size', 'SalesQuantity',
       'SalesDollars', 'SalesPrice', 'SalesDate', 'Volume', 'Classification',
       'ExciseTax', 'VendorNo', 'VendorName'],
      dtype='str')

In [23]:
query = """
SELECT 
    "VendorNo",
    "Brand",
    SUM("SalesDollars") as "TotalSalesDollars",
    SUM("SalesPrice") as "TotalSalesPrice",
    SUM("SalesQuantity") as "TotalSalesQuantity",
    SUM("ExciseTax") as "TotalExciseTax"
FROM sales
GROUP BY
    "VendorNo",
    "Brand"
ORDER BY
    "TotalSalesDollars"
"""

sales_summary = pd.read_sql(query, conn)
sales_summary

VendorNo  Brand  TotalSalesDollars  TotalSalesPrice  \
0          8004   5287               0.98             0.98   
1          9206   2773               0.99             0.99   
2          3924   9123               1.98             0.99   
3          3252   3933               1.98             0.99   
4         10050   3623               1.98             1.98   
...         ...    ...                ...              ...   
11267      3960   3545         4223107.62        545778.28   
11268      3960   4261         4475972.88        420050.01   
11269     17035   8068         4538120.60        461140.15   
11270      4425   3405         4819073.49        561512.37   
11271      1128   1233         5101919.51        672819.31   

       TotalSalesQuantity  TotalExciseTax  
0                     2.0            0.10  
1                     1.0            0.05  
2                     2.0            0.10  
3                     2.0            0.10  
4                     2.0            0.10  
...                   ...             ...  
11267            135838.0       249587.83  
11268            200412.0       368242.80  
11269            187140.0       343854.07  
11270            160247.0       294438.66  
11271            142049.0       260999.20  

[11272 rows x 6 columns]

In [24]:
import time

# Start the timer
start = time.time()

query = """
SELECT 
    pp."VendorNumber",
    pp."Brand",
    pp."Price" AS "ActualPrice",
    pp."PurchasePrice",
    SUM(s."SalesQuantity") AS "TotalSalesQuantity",
    SUM(s."SalesDollars") AS "TotalSalesDollars",
    SUM(s."SalesPrice") AS "TotalSalesPrice",
    SUM(s."ExciseTax") AS "TotalExciseTax",
    SUM(vi."Quantity") AS "TotalPurchaseQuantity",
    SUM(vi."Dollars") AS "TotalPurchaseDollars",
    SUM(vi."Freight") AS "TotalFreightCost"
FROM "purchase_prices" pp
JOIN "sales" s
    ON pp."VendorNumber" = s."VendorNo"
    AND pp."Brand" = s."Brand"
JOIN "vendor_invoice" vi
    ON pp."VendorNumber" = vi."VendorNumber"
GROUP BY 
    pp."VendorNumber", 
    pp."Brand", 
    pp."Price", 
    pp."PurchasePrice"
"""

final_table = pd.read_sql_query(query, conn)

# End the timer
end = time.time()

print(f"Execution Time: {round(end - start, 2)} seconds")
final_table

Execution Time: 144.83 seconds


VendorNumber  Brand  ActualPrice  PurchasePrice  TotalSalesQuantity  \
0                 2  90085        36.99          23.86                72.0   
1                 2  90609        24.99          17.00                96.0   
2                60    771        14.99          11.44              1645.0   
3                60   3979        20.99          16.27            137585.0   
4               105   2529        29.99          23.25               480.0   
...             ...    ...          ...            ...                 ...   
11217        172662   4215        25.99          19.40              3741.0   
11218        173357   2804        44.99          32.14              6300.0   
11219        173357   3666        24.99          18.79             16200.0   
11220        173357   3848        30.99          23.30               270.0   
11221        173357   3909        24.99          19.37             44190.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  \
0           2.663280e+03     1.183680e+03            8.00   
1           2.399040e+03     1.799280e+03            2.08   
2           2.465855e+04     1.731345e+04         1295.35   
3           2.340509e+06     1.458888e+06       252842.10   
4           1.439520e+04     2.399200e+03          377.60   
...                  ...              ...             ...   
11217       9.722859e+04     7.838584e+04         2950.75   
11218       2.834370e+05     1.437430e+05         4964.85   
11219       4.048380e+05     2.192873e+05         6353.55   
11220       8.367300e+03     4.183650e+03          211.95   
11221       1.104308e+06     6.511144e+05        34824.15   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalFreightCost  
0                     2624.0          4.504704e+04            216.64  
1                     5904.0          1.013558e+05            487.44  
2                   156156.0          2.533418e+06          12128.16  
3                 11588668.0          1.880103e+08         900056.48  
4                      664.0          2.341240e+04            124.78  
...                      ...                   ...               ...  
11217               169416.0          3.609635e+06          18547.36  
11218               141290.0          2.913587e+06          14377.50  
11219               388050.0          8.002106e+06          39487.50  
11220                 5970.0          1.231093e+05            607.50  
11221              1152210.0          2.376010e+07         117247.50  

[11222 rows x 11 columns]

In [25]:
### Optimized Way to run just above query

import time

# Start the timer
start = time.time()
query = """
WITH "FreightSummary" AS (
    SELECT
        "VendorNumber",
        SUM("Freight") AS "FreightCost"
    FROM "vendor_invoice"
    GROUP BY "VendorNumber"
),
"PurchaseSummary" AS (
    SELECT
        p."VendorNumber",
        p."VendorName",
        p."Brand",
        p."Description",
        p."PurchasePrice",
        pp."Price" AS "ActualPrice",
        pp."Volume",
        SUM(p."Quantity") AS "TotalPurchaseQuantity",
        SUM(p."Dollars") AS "TotalPurchaseDollars"
    FROM "purchases" p
    JOIN "purchase_prices" pp 
        ON p."Brand" = pp."Brand"
    WHERE p."PurchasePrice" > 0
    GROUP BY 
        p."VendorNumber", 
        p."VendorName", 
        p."Brand", 
        p."Description", 
        p."PurchasePrice", 
        pp."Price", 
        pp."Volume"
),
"SalesSummary" AS (
    SELECT
        "VendorNo",
        "Brand",
        SUM("SalesQuantity") AS "TotalSalesQuantity",
        SUM("SalesDollars") AS "TotalSalesDollars",
        SUM("SalesPrice") AS "TotalSalesPrice",
        SUM("ExciseTax") AS "TotalExciseTax"
    FROM "sales"
    GROUP BY "VendorNo", "Brand"
)
SELECT
    ps."VendorNumber",
    ps."VendorName",
    ps."Brand",
    ps."Description",
    ps."PurchasePrice",
    ps."ActualPrice",
    ps."Volume",
    ps."TotalPurchaseQuantity",
    ps."TotalPurchaseDollars",
    ss."TotalSalesQuantity",
    ss."TotalSalesDollars",
    ss."TotalSalesPrice",
    ss."TotalExciseTax",
    fs."FreightCost"
FROM "PurchaseSummary" ps
LEFT JOIN "SalesSummary" ss
    ON ps."VendorNumber" = ss."VendorNo"
    AND ps."Brand" = ss."Brand"
LEFT JOIN "FreightSummary" fs
    ON ps."VendorNumber" = fs."VendorNumber"
ORDER BY ps."TotalPurchaseDollars" DESC
"""

vendor_sales_summary = pd.read_sql_query(query, conn)

# End the timer
end = time.time()

print(f"Execution Time: {round(end - start, 2)} seconds")
vendor_sales_summary

Execution Time: 5.34 seconds


VendorNumber                   VendorName  Brand  \
0              1128  BROWN-FORMAN CORP             1233   
1              4425        MARTIGNETTI COMPANIES   3405   
2             17035  PERNOD RICARD USA             8068   
3              3960  DIAGEO NORTH AMERICA INC      4261   
4              3960  DIAGEO NORTH AMERICA INC      3545   
...             ...                          ...    ...   
10687          9815  WINE GROUP INC                8527   
10688          8004  SAZERAC CO INC                5683   
10689          3924  HEAVEN HILL DISTILLERIES      9123   
10690          3960  DIAGEO NORTH AMERICA INC      6127   
10691          7245  PROXIMO SPIRITS INC.          3065   

                       Description  PurchasePrice  ActualPrice Volume  \
0          Jack Daniels No 7 Black          26.27        36.99   1750   
1            Tito's Handmade Vodka          23.19        28.99   1750   
2                 Absolut 80 Proof          18.24        24.99   1750   
3           Capt Morgan Spiced Rum          16.17        22.99   1750   
4                  Ketel One Vodka          21.89        29.99   1750   
...                            ...            ...          ...    ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99    750   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49     50   
10689              Deep Eddy Vodka           0.74         0.99     50   
10690  The Club Strawbry Margarita           1.47         1.99    200   
10691     Three Olives Grape Vodka           0.71         0.99     50   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                   145080.0            3811251.60            142049.0   
1                   164038.0            3804041.22            160247.0   
2                   187407.0            3418303.68            187140.0   
3                   201682.0            3261197.94            200412.0   
4                   138109.0            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                    2.0                  2.64                 5.0   
10688                    6.0                  2.34               134.0   
10689                    2.0                  1.48                 2.0   
10690                    1.0                  1.47                72.0   
10691                    1.0                  0.71                86.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  
0             5101919.51        672819.31       260999.20     68601.68  
1             4819073.49        561512.37       294438.66    144929.24  
2             4538120.60        461140.15       343854.07    123780.22  
3             4475972.88        420050.01       368242.80    257032.07  
4             4223107.62        545778.28       249587.83    257032.07  
...                  ...              ...             ...          ...  
10687              15.95            10.96            0.55     27100.41  
10688              65.66             1.47            7.04     50293.62  
10689               1.98             0.99            0.10     14069.87  
10690             143.28            77.61           15.12    257032.07  
10691              85.14            33.66            4.46     38994.78  

[10692 rows x 14 columns]

This query generates a vendor-wise sales and purchase summary, which is valuable for:

**Performance Optimization:**
- The query involves heavy joins and aggregations on large datasets like sales and purchases.
- Storing the pre-aggregated results avoids repeated expensive computations.
- Helps in analyzing sales, purchases, and pricing for different vendors and brands.
- Future Benefits of Storing this data for faster Dashboarding & Reporting.
- Instead of running expensive queries each time, dashboards can fetch data quickly from vendor_sales_summary.

In [36]:
vendor_sales_summary.dtypes
# volume is str convert float64

VendorNumber               int64
VendorName                   str
Brand                      int64
Description                  str
PurchasePrice            float64
ActualPrice              float64
Volume                   float64
TotalPurchaseQuantity    float64
TotalPurchaseDollars     float64
TotalSalesQuantity       float64
TotalSalesDollars        float64
TotalSalesPrice          float64
TotalExciseTax           float64
FreightCost              float64
GrossProfit              float64
ProfitMargin             float64
StockTurnover            float64
SalestoPurchaseRatio     float64
dtype: object

In [27]:
vendor_sales_summary.isnull().sum()
# missing value present 

VendorNumber               0
VendorName                 0
Brand                      0
Description                0
PurchasePrice              0
ActualPrice                0
Volume                     0
TotalPurchaseQuantity      0
TotalPurchaseDollars       0
TotalSalesQuantity       178
TotalSalesDollars        178
TotalSalesPrice          178
TotalExciseTax           178
FreightCost                0
dtype: int64

In [28]:
vendor_sales_summary['VendorName'].unique()
# inconsistent space are persent

<StringArray>
['BROWN-FORMAN CORP          ',       'MARTIGNETTI COMPANIES',
 'PERNOD RICARD USA          ', 'DIAGEO NORTH AMERICA INC   ',
 'BACARDI USA INC            ', 'JIM BEAM BRANDS COMPANY    ',
 'MAJESTIC FINE WINES        ', 'ULTRA BEVERAGE COMPANY LLP ',
 'STOLI GROUP,(USA) LLC      ', 'PROXIMO SPIRITS INC.       ',
 ...
             'UNCORKED       ', 'BRONCO WINE COMPANY        ',
 'MILTONS DISTRIBUTING CO    ', 'TRUETT HURST               ',
 'LAUREATE IMPORTS CO        ', 'FANTASY FINE WINES CORP    ',
 'AAPER ALCOHOL & CHEMICAL CO', 'SILVER MOUNTAIN CIDERS     ',
 'CAPSTONE INTERNATIONAL     ', 'FLAVOR ESSENCE INC         ']
Length: 129, dtype: str

In [29]:
vendor_sales_summary['Volume'] = vendor_sales_summary ['Volume'].astype('float64')
vendor_sales_summary.fillna(0, inplace = True)
vendor_sales_summary['VendorName'] = vendor_sales_summary['VendorName'].str.strip()

In [30]:
vendor_sales_summary['GrossProfit'] = vendor_sales_summary['TotalSalesDollars'] - vendor_sales_summary['TotalPurchaseDollars']

In [31]:
vendor_sales_summary['ProfitMargin'] = (vendor_sales_summary['GrossProfit'] / vendor_sales_summary['TotalSalesDollars'])*100

In [32]:
vendor_sales_summary['StockTurnover'] = vendor_sales_summary['TotalSalesQuantity'] / vendor_sales_summary['TotalPurchaseQuantity']

In [33]:
vendor_sales_summary['SalestoPurchaseRatio'] = vendor_sales_summary['TotalSalesDollars'] / vendor_sales_summary['TotalPurchaseDollars']

In [34]:
vendor_sales_summary

VendorNumber                VendorName  Brand  \
0              1128         BROWN-FORMAN CORP   1233   
1              4425     MARTIGNETTI COMPANIES   3405   
2             17035         PERNOD RICARD USA   8068   
3              3960  DIAGEO NORTH AMERICA INC   4261   
4              3960  DIAGEO NORTH AMERICA INC   3545   
...             ...                       ...    ...   
10687          9815            WINE GROUP INC   8527   
10688          8004            SAZERAC CO INC   5683   
10689          3924  HEAVEN HILL DISTILLERIES   9123   
10690          3960  DIAGEO NORTH AMERICA INC   6127   
10691          7245      PROXIMO SPIRITS INC.   3065   

                       Description  PurchasePrice  ActualPrice  Volume  \
0          Jack Daniels No 7 Black          26.27        36.99  1750.0   
1            Tito's Handmade Vodka          23.19        28.99  1750.0   
2                 Absolut 80 Proof          18.24        24.99  1750.0   
3           Capt Morgan Spiced Rum          16.17        22.99  1750.0   
4                  Ketel One Vodka          21.89        29.99  1750.0   
...                            ...            ...          ...     ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99   750.0   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49    50.0   
10689              Deep Eddy Vodka           0.74         0.99    50.0   
10690  The Club Strawbry Margarita           1.47         1.99   200.0   
10691     Three Olives Grape Vodka           0.71         0.99    50.0   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                   145080.0            3811251.60            142049.0   
1                   164038.0            3804041.22            160247.0   
2                   187407.0            3418303.68            187140.0   
3                   201682.0            3261197.94            200412.0   
4                   138109.0            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                    2.0                  2.64                 5.0   
10688                    6.0                  2.34               134.0   
10689                    2.0                  1.48                 2.0   
10690                    1.0                  1.47                72.0   
10691                    1.0                  0.71                86.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  \
0             5101919.51        672819.31       260999.20     68601.68   
1             4819073.49        561512.37       294438.66    144929.24   
2             4538120.60        461140.15       343854.07    123780.22   
3             4475972.88        420050.01       368242.80    257032.07   
4             4223107.62        545778.28       249587.83    257032.07   
...                  ...              ...             ...          ...   
10687              15.95            10.96            0.55     27100.41   
10688              65.66             1.47            7.04     50293.62   
10689               1.98             0.99            0.10     14069.87   
10690             143.28            77.61           15.12    257032.07   
10691              85.14            33.66            4.46     38994.78   

       GrossProfit  ProfitMargin  StockTurnover  SalestoPurchaseRatio  
0       1290667.91     25.297693       0.979108              1.338647  
1       1015032.27     21.062810       0.976890              1.266830  
2       1119816.92     24.675786       0.998575              1.327594  
3       1214774.94     27.139908       0.993703              1.372493  
4       1199901.61     28.412764       0.983556              1.396897  
...            ...           ...            ...                   ...  
10687        13.31     83.448276       2.500000              6.041667  
10688        63.32     96.436186      22.333333             28.059829  
10689         0.50 

In [38]:

db_conn = conn.raw_connection()  # conn is your SQLAlchemy engine
cursor = db_conn.cursor()

In [39]:
cursor.execute("""
CREATE TABLE IF NOT EXISTS "vendor_sales_summary" (
    "VendorNumber" INT,
    "VendorName" VARCHAR(255),
    "Brand" INT,
    "Description" VARCHAR(255),
    "PurchasePrice" DECIMAL(15,2),
    "ActualPrice" DECIMAL(15,2),
    "Volume" DECIMAL(15,2),
    "TotalPurchaseQuantity" INT,
    "TotalPurchaseDollars" DECIMAL(15,2),
    "TotalSalesQuantity" INT,
    "TotalSalesDollars" DECIMAL(15,2),
    "TotalSalesPrice" DECIMAL(15,2),
    "TotalExciseTax" DECIMAL(15,2),
    "FreightCost" DECIMAL(15,2),
    "GrossProfit" DECIMAL(15,2),
    "ProfitMargin" DECIMAL(15,2),
    "StockTurnover" DECIMAL(15,2),
    "SalesToPurchaseRatio" DECIMAL(15,2),
    PRIMARY KEY ("VendorNumber", "Brand")
);
""")


In [41]:
db_conn.commit()

In [45]:
pd.read_sql_query('Select * FROM vendor_sales_summary', conn)

VendorNumber                VendorName  Brand  \
0              1128         BROWN-FORMAN CORP   1233   
1              4425     MARTIGNETTI COMPANIES   3405   
2             17035         PERNOD RICARD USA   8068   
3              3960  DIAGEO NORTH AMERICA INC   4261   
4              3960  DIAGEO NORTH AMERICA INC   3545   
...             ...                       ...    ...   
10687          9815            WINE GROUP INC   8527   
10688          8004            SAZERAC CO INC   5683   
10689          3924  HEAVEN HILL DISTILLERIES   9123   
10690          3960  DIAGEO NORTH AMERICA INC   6127   
10691          7245      PROXIMO SPIRITS INC.   3065   

                       Description  PurchasePrice  ActualPrice  Volume  \
0          Jack Daniels No 7 Black          26.27        36.99  1750.0   
1            Tito's Handmade Vodka          23.19        28.99  1750.0   
2                 Absolut 80 Proof          18.24        24.99  1750.0   
3           Capt Morgan Spiced Rum          16.17        22.99  1750.0   
4                  Ketel One Vodka          21.89        29.99  1750.0   
...                            ...            ...          ...     ...   
10687  Concannon Glen Ellen Wh Zin           1.32         4.99   750.0   
10688  Dr McGillicuddy's Apple Pie           0.39         0.49    50.0   
10689              Deep Eddy Vodka           0.74         0.99    50.0   
10690  The Club Strawbry Margarita           1.47         1.99   200.0   
10691     Three Olives Grape Vodka           0.71         0.99    50.0   

       TotalPurchaseQuantity  TotalPurchaseDollars  TotalSalesQuantity  \
0                   145080.0            3811251.60            142049.0   
1                   164038.0            3804041.22            160247.0   
2                   187407.0            3418303.68            187140.0   
3                   201682.0            3261197.94            200412.0   
4                   138109.0            3023206.01            135838.0   
...                      ...                   ...                 ...   
10687                    2.0                  2.64                 5.0   
10688                    6.0                  2.34               134.0   
10689                    2.0                  1.48                 2.0   
10690                    1.0                  1.47                72.0   
10691                    1.0                  0.71                86.0   

       TotalSalesDollars  TotalSalesPrice  TotalExciseTax  FreightCost  \
0             5101919.51        672819.31       260999.20     68601.68   
1             4819073.49        561512.37       294438.66    144929.24   
2             4538120.60        461140.15       343854.07    123780.22   
3             4475972.88        420050.01       368242.80    257032.07   
4             4223107.62        545778.28       249587.83    257032.07   
...                  ...              ...             ...          ...   
10687              15.95            10.96            0.55     27100.41   
10688              65.66             1.47            7.04     50293.62   
10689               1.98             0.99            0.10     14069.87   
10690             143.28            77.61           15.12    257032.07   
10691              85.14            33.66            4.46     38994.78   

       GrossProfit  ProfitMargin  StockTurnover  SalestoPurchaseRatio  
0       1290667.91     25.297693       0.979108              1.338647  
1       1015032.27     21.062810       0.976890              1.266830  
2       1119816.92     24.675786       0.998575              1.327594  
3       1214774.94     27.139908       0.993703              1.372493  
4       1199901.61     28.412764       0.983556              1.396897  
...            ...           ...            ...                   ...  
10687        13.31     83.448276       2.500000              6.041667  
10688        63.32     96.436186      22.333333             28.059829  
10689         0.50 

In [ ]:

vendor_sales_summary.to_sql('vendor_sales_summary', conn, if_exists='replace', index=False)

692